### PR TITLE
[vcpkg] vcpkg_configure_make: Support macOS cross-compile

### PR DESCRIPTION
**Pull request by past-due**
**Describe the pull request**

Support macOS cross-compile in `vcpkg_configure_make` (ex. building an `arm64-osx` triplet on an x86_64 Mac).

- What does your PR fix?

Previously, ports like `gettext` that use `vcpkg_configure_make` were building successfully, but for the native architecture regardless of the osx triplet used.

So, for example, trying to build `gettext:arm64-osx` on a 64-bit Intel Mac would actually build an `x86_64` architecture library.

This sets the proper flags to ensure the libraries are built for the desired target architecture.
(i.e. Building `gettext:arm64-osx` on a 64-bit Intel Mac yields an `arm64` architecture library.)

- Which triplets are supported/not supported? Have you updated the CI baseline?

This should impact cross-compiling osx triplet scenarios.

- Additional notes

@Neumann-A:

I added all of this logic in `vcpkg_configure_make` because there are other consumers of `vcpkg_internal_get_cmake_vars` (like `vcpkg_configure_meson`) that I haven't had time to look into yet.

Theoretically, `-arch ${TARGET_ARCH} -isysroot ${VCPKG_DETECTED_CMAKE_OSX_SYSROOT}` could be added to CFLAGS / CXXFLAGS in `get_cmake_vars`, but, for example, meson appears to use some sort of "cross file" that specifies architectures, so I'm unsure whether it should get this addition to CFLAGS / CXXFLAGS?

### CODE REVIEW

**Comment by JackBoosY** [comment]
cc @Neumann-A for review this PR.

**Comment by Neumann-A** [comment]
> Theoretically, -arch ${TARGET_ARCH} -isysroot ${VCPKG_DETECTED_CMAKE_OSX_SYSROOT} could be added to CFLAGS / CXXFLAGS in get_cmake_vars, but, for example, meson appears to use some sort of "cross file" that specifies architectures, so I'm unsure whether it should get this addition to CFLAGS / CXXFLAGS?

we are running meson in `--plain` mode so it probably also requires those flags.

**Comment by Neumann-A** [comment]
An what happens if you add the settings to the compiler flags in the toolchain? Does CMake deduplicate the flags?

**Comment by past-due** [comment]
> An what happens if you add the settings to the compiler flags in the toolchain? Does CMake deduplicate the flags?

Just tested this, and CMake does _**not**_ de-duplicate the flags. Thus putting them in the toolchain isn't a satisfactory option.

> we are running meson in `--plain` mode so it probably also requires those flags.

Testing `fribidi:arm64-osx` (which uses `vcpkg_configure_meson`) yields the same issue as `gettext:arm64-osx` when building on an Intel64 Mac - it builds an x86_64 arch library.

So I've moved this compiler flags logic to `get_cmake_vars`. Additionally, I wrote it in a way that it should (theoretically) support "universal" (multi-arch) osx libraries, if and when the rest of vcpkg adds full support for this.

**Comment by past-due** [comment]
I don't believe the single CI failure is related to this PR?
> REGRESSION: mlpack:x86-windows. If expected, add mlpack:x86-windows=fail to .\scripts\ci.baseline.txt

File: scripts/cmake/vcpkg_configure_make.cmake [hunk 1]
```diff
@@ -402,6 +416,24 @@ function(vcpkg_configure_make)
...
+    if(VCPKG_TARGET_IS_OSX)
+        if (_csc_AUTOCONFIG AND NOT _csc_BUILD_TRIPLET OR _csc_DETERMINE_BUILD_TRIPLET)
+            _vcpkg_determine_autotools_host_arch_mac(BUILD_ARCH) # machine you are building on => --build=
+            _vcpkg_determine_autotools_target_arch_mac(TARGET_ARCH)
+            # --build: the machine you are building on
+            # --host: the machine you are building for
+            # --target: the machine that CC will produce binaries for
+            # https://stackoverflow.com/questions/21990021/how-to-determine-host-value-for-configure-when-using-cross-compiler
+            # Only for ports using autotools so we can assume that they follow the common conventions for build/target/host
+            set(_csc_BUILD_TRIPLET "--build=${BUILD_ARCH}-apple-darwin")
```
Comment (Neumann-A) [inline scripts/cmake/vcpkg_configure_make.cmake:429]: have you checked the name against `config.guess`?

Comment (past-due) [reply scripts/cmake/vcpkg_configure_make.cmake:435]: @Neumann-A: `config.guess` uses `<arch>-apple-darwin`.

Although it doesn't impact the actual outcome, I'd added `config.guess`'s exception for mapping `arm64` -> `aarch64` for completeness. 👍

**Comment by past-due** [comment]
@NancyLi1013 @JackBoosY The `gettext` failure on Windows appears to be also happening on the master branch: https://github.com/microsoft/vcpkg/runs/1716299865

**Comment by JackBoosY** [comment]
gettext does have regression on master, but the reason is different from this:
```
-- Building arm64-windows-dbg
CMake Error at scripts/cmake/vcpkg_build_make.cmake:205 (message):
  libtool could not find a file being linked against!
Call Stack (most recent call first):
  scripts/cmake/vcpkg_install_make.cmake:26 (vcpkg_build_make)
  ports/gettext/portfile.cmake:59 (vcpkg_install_make)
  scripts/ports.cmake:137 (include)
```

**Comment by Neumann-A** [comment]
The real error is:
```
*** Warning: linker path does not have real file for library -liconv.
*** I have the capability to make that library automatically link in when
*** you link to this library.  But I can only do this if you have a
*** shared version of the library, which you do not appear to have
*** because I did check the linker path looking for a file starting
*** with iconv and none of the candidates passed a file format test
*** using a file magic. Last file checked: /usr/bin/iconv.exe
*** The inter-library dependencies that have been dropped here will be
*** automatically added whenever a program is linked with this library
*** or is declared to -dlopen it.

*** Since this library must not contain undefined symbols,
*** because either the platform does not support them or
*** it was explicitly requested with -no-undefined,
*** libtool will only create a static version of it.
```

**Comment by past-due** [comment]
@Neumann-A @JackBoosY I can reproduce this same exact error with a fresh checkout of latest `master`, attempting to build `gettext:arm64-windows` (on an x64 PC).

A git bisect reveals that:
3ddaeb41800d263b5cffeef1c87b23bdc1f3b98f - works
af3c99bc65d46e77900ae4e826866f134dc903ae - breaks `gettext:arm64-windows`

**Comment by JackBoosY** [comment]
Waiting for #15740.

**Comment by past-due** [comment]
@JackBoosY Anything else needed to get this merged? It's now working well on CI (and in my tests).

(I have some follow-up PRs that depend on this one to fix specific ports.)

**Review by JackBoosY — APPROVED** [review]

**Comment by JackBoosY** [comment]
Should be good to me.

**Comment by past-due** [comment]
@dan-shaw: Please let me know if you have any questions that need to be answered before this can be merged. (With the recent conversion of the `harfbuzz`port  to meson (cc093a822431437b4a78945032f2996efe668a7d), many more ports are now broken on `arm64-osx`, and this PR is a pre-requisite for a future PR to fix meson builds on `arm64-osx`.)

File: scripts/cmake/vcpkg_configure_make.cmake [hunk 2]
```diff
@@ -402,6 +422,24 @@ function(vcpkg_configure_make)
...
+            if(NOT TARGET_ARCH MATCHES "${BUILD_ARCH}") # we don't need to specify the additional flags if we build natively.
```
Comment (BillyONeal) [inline scripts/cmake/vcpkg_configure_make.cmake:436]:
```suggestion
            if(NOT "${TARGET_ARCH}" STREQUAL "${BUILD_ARCH}") # we don't need to specify the additional flags if we build natively.

```

File: scripts/get_cmake_vars/CMakeLists.txt [hunk 3]
```diff
@@ -81,17 +87,30 @@ foreach(_env IN LISTS VCPKG_ENV_VARS_TO_CHECK)
...
 endforeach()
 
+macro(_vcpkg_adjust_flags flag_var)
+    if(MSVC) # Transform MSVC /flags to -flags due to bash scripts intepreting /flag as a path.
+        string(REGEX REPLACE "(^| )/" "\\1-" ${flag_var} "${${flag_var}}")
+    endif()
+    if(CMAKE_SYSTEM_NAME MATCHES "Darwin")
+        if("${flag_var}" IN_LIST VCPKG_LANG_FLAGS)
+            # macOS - append arch and isysroot if cross-compiling
+            if(NOT CMAKE_OSX_ARCHITECTURES MATCHES "${CMAKE_HOST_SYSTEM_PROCESSOR}")
```
Comment (BillyONeal) [inline scripts/get_cmake_vars/CMakeLists.txt:97]:
```suggestion
            if(NOT "${CMAKE_OSX_ARCHITECTURES}" STREQUAL "${CMAKE_HOST_SYSTEM_PROCESSOR}")

```

**Review by dan-shaw — APPROVED** [review]

**Comment by dan-shaw** [comment]
LGTM, ready to merge once CI is passing

**Comment by past-due** [comment]
> LGTM, ready to merge once CI is passing

@dan-shaw: Looks like the remaining CI failure is unrelated to this PR. (Seems a version info update was missed in 55466813897a75aa443c5dfe6e32fb66c56847d1.)

**Comment by JackBoosY** [comment]
Regression will be fixed in https://github.com/microsoft/vcpkg/pull/15933.

**Pull request merged**